### DISCUSSION TITLE
StoredProcedures / Web API Actions in Sql4CDS

### DISCUSSION THREAD

**Discussion by PRusakPL** [MarkMpn/Sql4Cds#349] [Q&A] [answered]
Hi @MarkMpn,

Can you tell how you are retrieving list of Web API available Functions / Actions (Sql4CDS nomenclature: Stored Procedures)?
I am missing e.g. GrantAccess action (why ?) Are there any restrictions in showing / enabling actions from DV in your tool?

BR
Piotr Rusak

**Comment by MarkMpn** [comment, marked as answer]
The code to retrieve the list of actions is at https://github.com/MarkMpn/Sql4Cds/blob/1ddcc9936a194ac57617fecb3e1936f008e133c2/MarkMpn.Sql4Cds.Engine/MessageCache.cs#L54

It will retrieve all the available actions, but not all are made available to invoke as stored procedures or table valued functions.

In order to be executable as a table valued function, a message must have:

1. request fields only of scalar values (except for a single PagingInfo parameter if the output is an entity collection)
2. response fields that are EITHER:
  a. A single field of an entity-derived type, OR
  b. A single field of an entity collection, OR
  c. one or more fields of a scalar type

To be executable as a stored procedure the same conditions apply, except that messages with no response fields are also acceptable.

In the case of the `GrantAccess` message, the input parameter is a complex type and therefore can't be represented using standard SQL values.

**Reply by PRusakPL** [reply]
Thank you @MarkMpn - it was really fast. Kudos for what you are sharing with community - You Rock Man !!
